buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -1881,35 +1892,25 @@ void transferTo(ByteBuf dst) {
         }
 
         ByteBuf slice() {
-            return slice != null ? slice : (slice = buf.slice(idx(offset), length()));
+            return slice != null ? slice : (slice = srcBuf.slice(srcIdx(offset), length()));
         }
 
         ByteBuf duplicate() {
-            return buf.duplicate().setIndex(idx(offset), idx(endOffset));
+            // We can only use the cached slice safely if it's equal to the source buffer,
+            // otherwise it might not be fully visible to us
+            return (slice == srcBuf ? srcBuf : srcBuf.slice()).duplicate();
```

Comment:
Why do we need to first slice and then also duplicate ?

Reply:
@normanmaurer oops, I had meant this to be `.slice(srcIdx(offset), length())`. But thinking some more and comparing further to the "original" impl I think you are right that `srcBuf.duplicate()` would be ok and obviously preferable overhead-wise.

I was concerned about the possibility of someone independently modifying the indices of a buffer after it was added, but we make it clear that the `CompositeByteBuf` takes "ownership" of all added buffers (mainly refcount-wise but I think indices would be assumed too).